new article Gérer les primitives et les references en javascript (#19)
lelionvert/a-software-crafter-s-journey: articles/2021/05/gerer-les-primitives-et-les-references-en-javascript.md

---
title: Gerer les primitives et les reférences en javascript
author: Ghassen ASKRI
published_date: 03 Mai 2021
description: Cet article a pour objectif de comprendre la différence entre ces deux variables en javascript
---

> ##Gérer les primitives et les références en JavaScript 

Reviewer: Attention, il me semble que pour que les titres s'affichent correctement en markdown, il te faut mettre un espaces entre les # et le texte.
```suggestion
## Gérer les primitives et les références en JavaScript 
```

Généralement la plupart des développeurs ne font pas attention lorsqu'ils déclarent leurs variables en Javascript et comme tout autre langage, deux classes se distinguent : les primitives et les références. 

Cet article a pour objectif de comprendre la différence entre ces deux variables.  

Avant de commencer, étudions les deux types de mémoire afin de comprendre comment faire la différence entre les primitives et les références.

###Stack & Heap 
Les variables en JavaScript (et dans la plupart des autres langages de programmation) sont stockées à deux endroits : la Stack et la Heap.  

La Stack est généralement une région continue de mémoire allouant un contexte local pour chaque fonction en cours d'exécution. Cette dernière est la plupart du temps, la plus protégée et la plus rapide. 

La Heap, quant à elle, est une région beaucoup plus grande où tout est allouée dynamiquement. Cette séparation est utile pour rendre l'exécution plus sûre face à la corruption.  

![title](images/pic-01.png)

###Gérer des Primitives  

La figure ci-dessus montre que toutes les primitives sont stockées avec leurs valeurs dans la Stack.  

Mettons à présent cela en pratique : 
```javascript

> let firstPrimitive = 9;

Reviewer: Il vaut mieux utiliser `const` que `let`.
Pour mes articles, généralement, je code avec des `var`, ce qui permet de prendre le bout de code et de le copier-coller directement dans la console du browser: l'avantage de `var`, c'est que tu peux copier-coller plusieurs fois le même bout de code sans que ça plante :)
C'est mon seul cas d'utilisation de `var`, sinon j'utilise uniquement `const` dans mes programmes professionnels (rarement `let`)

Après toi, tu as mis des liens vers de jsfiddle (c'est surement plus pratique quand on est sur mobile ^^)

Du coup je peux te conseiller cet article : https://github.com/udebella/blog/blob/master/dev/javascript/1-variable-declaration.md

let secondPrimitive  = firstPrimitive ;

> console.log('the second primitive equal to : '+ secondPrimitive);

Reviewer: Tu peux utiliser les templates strings :)
```suggestion
console.log(`the second primitive equal to : ${secondPrimitive}`);
```

Pour avoir accès à notre objet, il faut pointer sur sa référence. Au niveau de la Stack, il existe une copie par référence et les deux objets pointent sur la même référence. Cela entraine un changement de valeur dans le premier objet si le deuxième objet a été modifié et vice-versa puisque les deux objets partagent la même référence.  
```javascript
let firstReference = {lat:-34.397, lng:150.644};

let secondReference = firstReference ; 

secondReference.lng = 0.000;

console.log('secondLng ' + secondReference.lng);
console.log('firstLng  ' + firstReference.lng);
```
https://jsfiddle.net/fan7vxqk/  

On peut conclure que la manipulation des références exige une attention particulière pour ne pas modifier ou muter l’objet de départ.  

Lorsque l’on manipule des services REST, on gère des objets et on manipule des états sans parfois s’apercevoir qu’une action pourrait mettre en danger les états de l’application. 

 

Deux solutions classiques s’offrent à nous pour ne pas utiliser la même référence d'un objet :  

- Créer un nouvel objet qui comporte les mêmes attributs et les mêmes valeurs que le premier objet.  

> - Faire une copie d'un objet par une simple opération d'affectation tout en dissociant le couplage entre les deux objets. Javascript offre une méthode qui permet de faire cette manipulation : "Object.assign()" 

Reviewer: Il existe aussi l'operateur [rest properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) : 
```javascript
const secondeReference = {...firstReference}
```

Reviewer: L'inconvenient de `Object.assign` c'est que si tu te trompe d'ordre des paramètres ça fait pas du tout ce que tu t'attendais à avoir ^^

secondReference.lng = 0.000;

secondReference.pays.push("peche");

console.log('secondLng ', secondReference.lng);
console.log('firstLng ', firstReference.lng);
console.log('scondFruitsReference', secondReference.pays); 
console.log('firstFruitsReference', secondReference.pays); 
``` 
https://jsfiddle.net/fan7vxqk/2/ 

Dans l'exemple ci-dessus, nous avons cherché à construire un premier objet "firstReference". Celui-ci a ensuite été copié dans un deuxième objet "secondReference" en utilisant la méthode assign(). 

Nous avons ensuite modifié les deux attributs en passant par le deuxième objet “secondReference”.  

Cette modification n’a pas eu d’impact sur l’attribut “lng” qui est resté inchangé tandis que l’attribut fruits a été impacté et sa valeur a été modifié. 

On peut conclure que la méthode assign() ne pourra pas gérer l'imbrication des références tout simplement parce que l'attribut "pays" est de type tableau (référence).  

La méthode assign() ne présente pas la meilleure solution car elle ne gère pas l'imbrication des références.

Pour faire une copie complète et gérer l’imbrication, il y a trois solutions possibles:  

> - Faire un clonage superficiel en utilisant la standardisation ECMAScript 2018 const clone = {...original} Attention, l'expression n'est pas encore supportée par tous les navigateurs et pourrait engendrer des erreurs pour votre loader comme BabelLoader ! 

Reviewer: Le rest operator a exactement le même effet que `Object.assign`
```javascript
var firstReference = {lat:-34.397, lng:150.644, pays:['france','italie','allemagne']}
var second = {...firstReference}
second.pays.push('russie')
console.log(firstReference.pays) // ['france','italie','allemagne', 'russie']
```

Reviewer: La meilleure solution à mon avis c'est de ne jamais utiliser la fonction `push` parce qu'elle fait muter les variables.
```javascript
var firstReference = {lat:-34.397, lng:150.644, pays:['france','italie','allemagne']}
var second = {
    ...firstReference,
    pays: [...firstReference.pays, 'russie']
}
console.log(firstReference.pays) // ['france','italie','allemagne', 'russie']
```

Bon par contre ça marche tant qu'il n'y pas trop de niveau d'imbrication. Au delà de 2 ça commence à devenir horrible à utiliser...

---
title: Gerer les primitives et les reférences en javascript
author: Ghassen ASKRI
published_date: 03 Mai 2021
description: Cet article a pour objectif de comprendre la différence entre ces deux variables en javascript
---
# Gérer les primitives et les références en JavaScript 

Lors de l'écriture des programmes, il est souvent nécessaire d'avoir des variables pour nous permettre de stocker des valeurs.  

> Généralement la plupart des développeurs ne font pas attention lorsqu'ils déclarent leurs variables en Javascript et comme tout autre langage, deux classes se distinguent : les primitives et les références. 

Reviewer 2: > comme tout autre langage, deux classes se distinguent

je ne suis pas certain que ce soit si universel

Cet article a pour objectif de comprendre la différence entre ces deux variables.  

Avant de commencer, étudions les deux types de mémoire afin de comprendre comment faire la différence entre les primitives et les références.

> ###Stack & Heap 

Reviewer 2: après un titre de premier niveau, il devrait y avoir un titre de deuxième niveau

```suggestion
## Stack & Heap 
```

> Les variables en JavaScript (et dans la plupart des autres langages de programmation) sont stockées à deux endroits : la Stack et la Heap.  

Reviewer 2: je ne savais pas qu'il y avait aussi cette séparation en JS :open_mouth: 

as-tu une source ?

> La Stack est généralement une région continue de mémoire allouant un contexte local pour chaque fonction en cours d'exécution. Cette dernière est la plupart du temps, la plus protégée et la plus rapide. 

Reviewer 2: > la plus protégée

protégée de quoi ? par rapport à quoi ? je ne comprend pas

La Heap, quant à elle, est une région beaucoup plus grande où tout est allouée dynamiquement. Cette séparation est utile pour rendre l'exécution plus sûre face à la corruption.  

> ![title](images/pic-01.png)

Reviewer 2: `title` sert à donner un titre à l'image, ça peut etre utile pour une personne qui ne peut pas la voir (problème réseau, aveugle ...)

je pense que ça serait mieux de mettre une vraie description, ou au pire de supprimer `title`

Reviewer 2: de plus dans l'image, il y a :
* `String`
* `Number`
* `Boolean`
ce sont des type objets :
```javascript
typeof new String("toto") === "object"
```

il ne faut pas confondre avec les types primitifs (en minuscule)

```javascript
typeof "toto" === "string"
```

mais qui sont quand meme des objets :

```javascript
"toto".toUpperCase() === "TOTO"
```

enfin, à ma connaissance `Undefined` et `Null` avec des majuscule, n'existent pas en JS

## Gérer des Primitives  

> La figure ci-dessus montre que toutes les primitives sont stockées avec leurs valeurs dans la Stack.  

Reviewer 2: je pense que le texte descriptif devrait directement suivre l'image, sans avoir le titre au milieu

Mettons à présent cela en pratique : 
```javascript

> var firstPrimitive = 9;

Reviewer 2: `var` peut etre considéré comme étant déprécié

je te conseille d'utiliser `const` par défaut et `let` si tu dois avoir une variable mutable

```
https://jsfiddle.net/yg85mwjr/ 

Dans le code au sein du lien ci-dessus, une opération d'affectation sur les primitives permet de faire une copie de valeur au niveau de la Stack.

![title](images/pic-02.png)
Maintenant si l’on change la valeur de la première primitive "firstPrimitive", la deuxième primitive sera-t-elle affectée ? Rien ne change puisqu'il existe une copie par valeur au niveau de la Stack.  

```javascript
var firstPrimitive = 9;

var secondPrimitive  = firstPrimitive ; 

console.log('the second primitive equal to : '+ secondPrimitive);

firstPrimitive = 7;

console.log('the first primitive equal to : ' +firstPrimitive);
```

https://jsfiddle.net/92n3bejf/

On peut alors en conclure que lorsqu'il s'agit d'une primitive, une opération d'affectation engendrera une copie par valeur au niveau de la Stack.  

La plupart des développeurs junior savent qu'une simple affectation ne peut rien faire sur la valeur d'une variable. Malheureusement, à ce niveau-là, les problèmes techniques commencent à apparaître car ce comportement au niveau de la Stack n'est valable qu'avec les primitives et non avec les références.

> ###Gérer des références  

Reviewer 2: en markdown il est nécessaire d'avoir un espace après les `#` pour afficher un titre

Commençons par une déclaration simple de deux références. Nous découvrirons ensuite l’impact sur la Stack et nous pourrons mieux comprendre les manipulations.  

```javascript
var firstReference = {lat:-34.397, lng:150.644};  

var secondReference = firstReference ;  
```

> ![title](images/pic-03.png)

Reviewer 2: dans l'image il y a des choses qui ne correspondent pas au code

je ne sais pas ce que je suis censé en comprendre

>  On remarque que l’objet “secondReference” a été enregistré dans la mémoire nommée "Heap" avec la même référence que l’objet “firstRefrence” 

Reviewer 2: je suggère de marquer les mots techniques avec des backquotes

```suggestion
 On remarque que l’objet `secondReference` a été enregistré dans la mémoire nommée "Heap" avec la même référence que l’objet `firstRefrence”`
```

> Pour avoir accès à notre objet, il faut pointer sur sa référence. Au niveau de la Stack, il existe une copie par référence et les deux objets pointent sur la même référence. Cela entraine un changement de valeur dans le premier objet si le deuxième objet a été modifié et vice-versa puisque les deux objets partagent la même référence.  

Reviewer 2: > Au niveau de la Stack, il existe une copie par référence et les deux objets pointent sur la même référence.

si je me met à la place d'une personne qui n'a jamais fait de C++, je ne comprendrais peut etre pas "copie par référence"

c'est vrai en C++, je ne sais pas si c'est ce qui est fait en JS (c'est un détail d'implem des moteurs JS, je ne sais pas si c'est vraiment fait comme ça)

```javascript
var firstReference = {lat:-34.397, lng:150.644};

var secondReference = firstReference ; 

secondReference.lng = 0.000;

console.log('secondLng ' + secondReference.lng);
console.log('firstLng  ' + firstReference.lng);
```
https://jsfiddle.net/fan7vxqk/  

On peut conclure que la manipulation des références exige une attention particulière pour ne pas modifier ou muter l’objet de départ.  

> Lorsque l’on manipule des services REST, on gère des objets et on manipule des états sans parfois s’apercevoir qu’une action pourrait mettre en danger les états de l’application. 

Reviewer 2: j'ai l'impression que parler d'un API REST est déconnecté du sujet actuel

 

Deux solutions classiques s’offrent à nous pour ne pas utiliser la même référence d'un objet :  

- Créer un nouvel objet qui comporte les mêmes attributs et les mêmes valeurs que le premier objet.  

> - Faire une copie d'un objet par une simple opération d'affectation tout en dissociant le couplage entre les deux objets. Javascript offre une méthode qui permet de faire cette manipulation : "Object.assign()" 

Reviewer 2: je ne vois pas la différence entre les 2 solutions

Nous aurons au final un nouvel objet ”secondReference”  au niveau de la mémoire Heap et une nouvelle référence différente de celle du "firstRefrence" au niveau de la Stack.  

```javascript
var firstReference = {lat:-34.397, lng:150.644};

var secondReference = Object.assign({},firstReference); 

firstReference.lng = 0.000;

> console.log('secondLng ' + secondReference.lng);
> console.log('firstLng ' + firstReference.lng);

Reviewer 2: je pense que ça pourrait pas mal d'avoir le résultat en commentaire (pour les gens qui ont du mal à suivre et on la flemme d'aller voir le jsfiddle)

```suggestion
console.log('secondLng ' + secondReference.lng); // 150.644
console.log('firstLng ' + firstReference.lng); // 0.0
```

    ```
    Ci-desous l'implémentation de notre exemple en utilisant le "Spread syntax"
    ```javascript
    var firstReference = {lat:-34.397, lng:150.644, countries:['france','italie','allemagne']}
    var secondReference = {
        ...firstReference,
        countries: [...firstReference.countries, 'russie']
    }
    console.log(firstReference.countries) // ['france','italie','allemagne']
    console.log(secondReference.countries) // ['france','italie','allemagne', 'russie']
    ```

    https://jsfiddle.net/4vjcsy5a/3/

    Attention,l'expression n'est pas encore supportée par tous les navigateurs et pourrait engendrer des erreurs pour votre loader comme BabelLoader ! 

2. Utiliser la librairie lodash.
    ```javascript
    import * as _ from 'lodash'; 
    myObjCopy = _.cloneDeep(myObj);  
    ```

3. Faire parcourir l'objet original.</br>
    Il ne faut jamais oublier que lorsque votre objet original ne contient aucune référence, la méthode Object.assign() est la plus pratique.</br>
    Attention, l'ordre des paramètres de la méthode "Object.assign()" doivent etre respectés.

### Références :
- Stack vs Heap: https://www.quora.com/What-is-the-dif...  

> - Object.assign(): https://developer.mozilla.org/en-US/d...  

Reviewer 2: les liens ne sont pas correctes
